docs/HowToFAQ/FAQFunctionByMethodProof.md

```diff
@@ -0,0 +1,89 @@
+---
+title: I can't prove the equivalence between the method part of a function by method and the function itself
+---
+
+## Question
+
+I can't prove the equivalence between the method part of a function by method and the function itself.
+It seems that my invariants can't be maintained by the loop, and I don't know how to prove them.
+
+## Answer
+
+Consider the simple problem of computing the sum of a sequence of integers.
+You might be tempted to immediately write the following equivalence,
+with your best guess of an invariant
+and a hint at the end to help Dafny realize that the result is correct.
+However you cannot prove the invariant you just wrote.
+
+```
+{% include_relative FAQFunctionByMethodProof1.dfy %}
+```
+
+Let's step back for a moment. Let's unroll the computation of the function on a sequence of 3 elements.
+The function will compute:
+
+`s[0] + (s[1] + (s[2] + 0))`
+
+But the loop will compute
+
+`((0 + s[0]) + s[1]) + s[2]`
+
+In the function, we add the first element to the sum of the remaining.
+In the by method, we add the last element to the sum of the initial terms.
+If it was not the case that the addition was associative and 0 was a neutral element, there would be no immediate way to prove that the two computations are equivalent.
+
+There are essentially three solutions around that:
+
+* Improve the invariant, knowing `+` is associative
+* Change the order of computation of the function so that it matches the by method
+* Change the order of computation of the by method so that it matches the function
+
```

Comment:
One can imagine a more "intrusive" solution: make the function tail-call, and remove the while loop altogether.